Add Unsafe Rust Deep Dive (#2806)
google/comprehensive-rust: src/unsafe-deep-dive/foundations/when-is-unsafe-used.md

# When is unsafe used?

The unsafe keyword indicates that the programmer is responsible for upholding
Rust's safety guarantees.

The keyword has two roles:

> - define pre-conditions must be satisfied

Reviewer: ```suggestion
- define pre-conditions that must be satisfied
```

> - verify that those defined pre-conditions are satisfied

Reviewer: ```suggestion
- assert to the compiler (= promise) that those defined pre-conditions are satisfied
```

## Further references

- [The unsafe keyword chapter of the Rust Reference](https://doc.rust-lang.org/reference/unsafe-keyword.html)

<details>

Places where pre-conditions can be defined (Role 1)

> - [unsafe functions] (`unsafe fn { ... }`). Example: `get_unchecked` method on

Reviewer: ```suggestion
- [unsafe functions] (`unsafe fn foo() { ... }`). Example: `get_unchecked` method on
```
